File: apps/analysis-worker/analysis_worker/repositories.py
from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime, timezone
from enum import Enum
from typing import Dict, Optional, Protocol


class JobStatus(str, Enum):
    QUEUED = "queued"
    RUNNING = "running"
    SUCCEEDED = "succeeded"
    FAILED = "failed"


@dataclass(frozen=True)
class JobRecord:
    job_id: str
    status: str
    artifact_paths: Dict[str, str] = field(default_factory=dict)
    error_message: Optional[str] = None


class JobRepository(Protocol):
    def get(self, job_id: str) -> Optional[JobRecord]:
        ...

    def mark_running(self, job_id: str) -> None:
        ...

    def mark_succeeded(self, job_id: str, artifact_paths: Dict[str, str]) -> None:
        ...

    def mark_failed(self, job_id: str, error_message: str) -> None:
        ...


class FirestoreJobRepository:
    def __init__(self, collection_name: str) -> None:
        from google.cloud import firestore

        self._firestore = firestore
        self._client = firestore.Client()
        self._collection = self._client.collection(collection_name)

    def get(self, job_id: str) -> Optional[JobRecord]:
        snapshot = self._collection.document(job_id).get()
        if not snapshot.exists:
            return None
        data = snapshot.to_dict() or {}
        return JobRecord(
            job_id=job_id,
            status=str(data.get("status", "")),
            artifact_paths=dict(data.get("artifact_paths") or data.get("artifactPaths") or {}),
            error_message=data.get("error_message") or data.get("errorMessage"),
        )

    def mark_running(self, job_id: str) -> None:
        self._collection.document(job_id).set(
            {
                "status": JobStatus.RUNNING.value,
                "updated_at": self._firestore.SERVER_TIMESTAMP,
                "error_message": None,
            },
            merge=True,
        )

    def mark_succeeded(self, job_id: str, artifact_paths: Dict[str, str]) -> None:
        self._collection.document(job_id).set(
            {
                "status": JobStatus.SUCCEEDED.value,
                "artifact_paths": artifact_paths,
                "updated_at": self._firestore.SERVER_TIMESTAMP,
                "error_message": None,
            },
            merge=True,
        )

    def mark_failed(self, job_id: str, error_message: str) -> None:
        self._collection.document(job_id).set(
            {
                "status": JobStatus.FAILED.value,
                "updated_at": self._firestore.SERVER_TIMESTAMP,
                "error_message": error_message,
            },
            merge=True,
        )


class InMemoryJobRepository:
    def __init__(self) -> None:
        self.records: Dict[str, JobRecord] = {}
        self.transitions: Dict[str, str] = {}

    def get(self, job_id: str) -> Optional[JobRecord]:
        return self.records.get(job_id)

    def mark_running(self, job_id: str) -> None:
        self.transitions[job_id] = JobStatus.RUNNING.value
        self.records[job_id] = JobRecord(job_id=job_id, status=JobStatus.RUNNING.value)

    def mark_succeeded(self, job_id: str, artifact_paths: Dict[str, str]) -> None:
        self.transitions[job_id] = JobStatus.SUCCEEDED.value
        self.records[job_id] = JobRecord(
            job_id=job_id,
            status=JobStatus.SUCCEEDED.value,
            artifact_paths=artifact_paths,
        )

    def mark_failed(self, job_id: str, error_message: str) -> None:
        self.transitions[job_id] = JobStatus.FAILED.value
        self.records[job_id] = JobRecord(
            job_id=job_id,
            status=JobStatus.FAILED.value,
            error_message=error_message,
        )


def utc_now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()

File: apps/analysis-worker/src/repositories.ts
import { Firestore, FieldValue, CollectionReference } from "@google-cloud/firestore";

export const JobStatus = {
  QUEUED: "queued",
  RUNNING: "running",
  SUCCEEDED: "succeeded",
  FAILED: "failed",
} as const;

export type JobStatusType = (typeof JobStatus)[keyof typeof JobStatus];

export interface JobRecord {
  jobId: string;
  status: string;
  artifactPaths: Record<string, string>;
  errorMessage: string | null;
}

export interface JobRepository {
  get(jobId: string): Promise<JobRecord | null>;
  markRunning(jobId: string): Promise<void>;
  markSucceeded(jobId: string, artifactPaths: Record<string, string>): Promise<void>;
  markFailed(jobId: string, errorMessage: string): Promise<void>;
}

export class FirestoreJobRepository implements JobRepository {
  private collection: CollectionReference;
  private firestoreFieldValue = FieldValue;

  constructor(collectionName: string) {
    this.collection = new Firestore().collection(collectionName);
  }

  async get(jobId: string): Promise<JobRecord | null> {
    const snapshot = await this.collection.doc(jobId).get();
    if (!snapshot.exists) {
      return null;
    }
    const data = snapshot.data() || {};
    return {
      jobId,
      status: String(data.status || ""),
      artifactPaths: { ...(data.artifact_paths || data.artifactPaths || {}) },
      errorMessage: data.error_message || data.errorMessage || null,
    };
  }

  async markRunning(jobId: string): Promise<void> {
    await this.collection.doc(jobId).set(
      {
        status: JobStatus.RUNNING,
        updated_at: this.firestoreFieldValue.serverTimestamp(),
        error_message: null,
      },
      { merge: true },
    );
  }

  async markSucceeded(jobId: string, artifactPaths: Record<string, string>): Promise<void> {
    await this.collection.doc(jobId).set(
      {
        status: JobStatus.SUCCEEDED,
        artifact_paths: artifactPaths,
        updated_at: this.firestoreFieldValue.serverTimestamp(),
        error_message: null,
      },
      { merge: true },
    );
  }

  async markFailed(jobId: string, errorMessage: string): Promise<void> {
    await this.collection.doc(jobId).set(
      {
        status: JobStatus.FAILED,
        updated_at: this.firestoreFieldValue.serverTimestamp(),
        error_message: errorMessage,
      },
      { merge: true },
    );
  }
}

export class InMemoryJobRepository implements JobRepository {
  public records: Record<string, JobRecord> = {};
  public transitions: Record<string, string> = {};

  async get(jobId: string): Promise<JobRecord | null> {
    return this.records[jobId] || null;
  }

  async markRunning(jobId: string): Promise<void> {
    this.transitions[jobId] = JobStatus.RUNNING;
    this.records[jobId] = { jobId, status: JobStatus.RUNNING, artifactPaths: {}, errorMessage: null };
  }

  async markSucceeded(jobId: string, artifactPaths: Record<string, string>): Promise<void> {
    this.transitions[jobId] = JobStatus.SUCCEEDED;
    this.records[jobId] = {
      jobId,
      status: JobStatus.SUCCEEDED,
      artifactPaths: { ...artifactPaths },
      errorMessage: null,
    };
  }

  async markFailed(jobId: string, errorMessage: string): Promise<void> {
    this.transitions[jobId] = JobStatus.FAILED;
    this.records[jobId] = {
      jobId,
      status: JobStatus.FAILED,
      artifactPaths: {},
      errorMessage,
    };
  }
}

export function utcNowIso(): string {
  return new Date().toISOString();
}
